syntax = "proto3";

package pythonserver;

// The service definition
service EventService {
  // Bidirectional streaming - both sides can send messages anytime
  rpc EventStream(stream ClientMessage) returns (stream ServerMessage);
}

// Messages from C# client to Python server
message ClientMessage {
  oneof message_type {
    AnalyticEvent analytic_event = 1;
    GetRecommendationsRequest get_recommendations = 2;
  }
}

message AnalyticEvent {
  string action = 1;
  string target = 2;
  map<string, string> metadata = 3;
}

message GetRecommendationsRequest {
  string request_id = 1;
  string user_id = 2;
}

// Messages from Python server to C# client
message ServerMessage {
  oneof message_type {
    RecommendationsResponse recommendations_response = 1;
    SaveStateRequest save_state = 2;
  }
}

message RecommendationsResponse {
  string request_id = 1;
  string user_id = 2;
  repeated string recommendations = 3;
}

message SaveStateRequest {
  int32 analytics_count = 1;
  map<string, string> user_preferences = 2;
}
